base/range.jl

```diff
@@ -1143,6 +1143,15 @@ function intersect(r::StepRange, s::StepRange)
     step(r) < zero(step(r)) ? StepRange{T,S}(n, -a, m) : StepRange{T,S}(m, a, n)
 end
 
+function intersect(r1::AbstractRange, r2::AbstractRange)
+    if length(r2) < length(r1)
+        r1, r2 = r2, r1
+    end
```

Comment:
I think this may cause type-stability problems, e.g. `@code_warntype intersect(1:10, 1:1.0:9)`. 

Even if I replace the swop with `return intersect_new(r2, r1)`, I still get `Body::Union{Vector{Float64}, Vector{Int64}}` for this example, due to differing eltypes.

Reply:
Maybe? The present behaviour BTW is this:
```
julia> intersect(1:1.0:9, 1:3)
3-element Vector{Float64}:
 1.0
 2.0
 3.0

julia> intersect(1:3, 1:1.0:9)
3-element Vector{Int64}:
 1
 2
 3
```
Are there situations where it might matter not to promote? That doesn't sound to me like it ought to be guaranteed by the verb `intersect`.